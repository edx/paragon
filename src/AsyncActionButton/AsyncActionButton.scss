.pgn__async-action-btn-icon {
  margin-right: .5em;  
}