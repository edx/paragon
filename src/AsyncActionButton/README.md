# Async Action Button

A button that can display different states. Each state has:

- A label (required)
- An icon
- an option to be disabled

These are controlled by the props `labels`, `icons`, and `disabledStates`.

Button states can be whatever you want them to be, but default icons exist for: 

- `pending`, `complete`, and `error`.

Control the state with the `state` prop. Example usage:

```jsx
<AsyncActionButton
  state="pending"
  labels={{
    default: 'Download',
    pending: 'Downloading',
    complete: 'Downloaded',
  }}
  icons={{
    default: <Icon className="fa fa-download" />,
    pending: <Icon className="fa fa-spinner fa-spin" />,
    complete: <Icon className="fa fa-check" />,
  }}
  disabledStates=['pending']
  className='btn-primary mr-2'
  ...
/>
```

All other props sent to this component will be passed through to the underlying Button component.