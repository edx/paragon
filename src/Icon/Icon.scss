.pgn__icon {
  display: block;
  height: 1.5rem; // 24px, Assumes base font size to be 16px
  position: relative;
  width: 1.5rem; // 24px, Assumes base font size to be 16px
  flex-shrink: 0;
}

.pgn__icon svg {
  display: block;
  height: 100%;
  position: relative;
  max-height: 100%;
  max-width: 100%;
  width: 100%;
}
