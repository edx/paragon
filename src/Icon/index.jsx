import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import newId from '../utils/newId';
import withDeprecatedProps, { DEPR_TYPES } from '../withDeprecatedProps';

function Icon({
  src: Component,
  id,
  className,
  hidden,
  screenReaderText,
  svgAttrs,
  ...attrs
}) {
  if (Component) {
    const mergedSvgProps = { ...svgAttrs, role: 'img' };
    // If no aria label is specified, hide this icon from screenreaders
    if (!svgAttrs['aria-label']) {
      mergedSvgProps['aria-label'] = '';
      mergedSvgProps['aria-hidden'] = true;
      mergedSvgProps.focusable = false;
    }
    return (
      <span
        className={classNames('pgn__icon', className)}
        id={id}
        {...attrs}
      >
        <Component {...mergedSvgProps} />
      </span>
    );
  }

  return (
    <>
      <span
        id={id || newId('Icon')}
        className={className}
        aria-hidden={hidden}
      />
      {screenReaderText && (
        <span className="sr-only">
          {screenReaderText}
        </span>
      )}
    </>
  );
}

Icon.propTypes = {
  // eslint-disable-next-line max-len
  /** An icon component to render. Example import of a Paragon icon component: `import { Check } from '@edx/paragon/dist/icon';` */
  src: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  /** HTML element attributes to pass through to the underlying svg element */
  svgAttrs: PropTypes.object, // eslint-disable-line
  // eslint-disable-next-line max-len
  /** the `id` property of the Icon element, by default this value is generated with the `newId` function with the `prefix` of `Icon`. */
  id: PropTypes.string,
  // eslint-disable-next-line max-len
  /** array of class names that will define what the Icon looks like. For example, a list of FontAwesome style names. */
  className: PropTypes.string,
  // eslint-disable-next-line max-len
  /** a boolean that determines the value of `aria-hidden` attribute on the Icon span, this value is `true` by default. */
  hidden: PropTypes.bool,
  // eslint-disable-next-line max-len
  /** a string or an element that will be used on a secondary span leveraging the `sr-only` style for screenreader only text, this value is `undefined` by default. This value is recommended for use unless the Icon is being used in a way that is purely decorative or provides no additional context for screen reader users. This field should be thought of the same way an `alt` attribute would be used for `image` tags.
   */
  screenReaderText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
};

Icon.defaultProps = {
  src: null,
  svgAttrs: {},
  id: undefined,
  hidden: true,
  screenReaderText: undefined,
  className: undefined,
};

export default withDeprecatedProps(Icon, 'Icon', {
  className: {
    deprType: DEPR_TYPES.FORMAT,
    expect: value => typeof value === 'string',
    transform: value => (Array.isArray(value) ? value.join(' ') : value),
    message: 'It should be a string.',
  },
});
