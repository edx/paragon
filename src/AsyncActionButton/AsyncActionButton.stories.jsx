import React from 'react';
import { storiesOf } from '@storybook/react';

import AsyncActionButton from './index';
import Icon from '../Icon';
import README from './README.md';

storiesOf('Basics|AsyncActionButton', module)
  .addParameters({ info: { text: README } })
  .add('basic usage', () => {
    const props = {
      labels: {
        default: 'Saved',
        pending: 'Saving',
        complete: 'Saved',
      },
      className: 'btn-primary mr-2',
    };
    return (
      <div>
        <AsyncActionButton {...props} />
        <AsyncActionButton state="pending" {...props} />
        <AsyncActionButton state="complete" {...props} />
      </div>
    );
  })
  .add('custom icons and disable during state', () => {
    const downloadButtonProps = {
      labels: {
        default: 'Download',
        pending: 'Downloading',
        complete: 'Downloaded',
      },
      icons: {
        default: <Icon className="fa fa-download" />,
        pending: <Icon className="fa fa-spinner fa-spin" />,
        complete: <Icon className="fa fa-check" />,
      },
      disabledStates: ['pending'],
      className: 'btn-primary mr-2',
    };
    return (
      <React.Fragment>
        <AsyncActionButton state="default" {...downloadButtonProps} />
        <AsyncActionButton state="pending" {...downloadButtonProps} />
        <AsyncActionButton state="complete" {...downloadButtonProps} />
      </React.Fragment>
    );
  })
  .add('custom states', () => {
    const buttonProps = {
      labels: {
        unedited: 'Save (no changes)',
        edited: 'Save Changes',
      },
      icons: {
        unedited: <Icon className="fa fa-save" />,
        edited: <Icon className="fa fa-save" />,
      },
      disabledStates: ['unedited'],
      className: 'btn-primary mr-2',
    };
    return (
      <React.Fragment>
        <AsyncActionButton state="unedited" {...buttonProps} />
        <AsyncActionButton state="edited" {...buttonProps} />
      </React.Fragment>
    );
  });
