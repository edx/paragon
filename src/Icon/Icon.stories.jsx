import React from 'react';
import { storiesOf } from '@storybook/react';

import FontAwesomeStyles from 'font-awesome/css/font-awesome.min.css';
import Icon from './index';
import README from './README.md';

storiesOf('Icon', module)
  .addParameters({ info: { text: README } })
  .add('basic usage', () => (
    <Icon
      className={[
        FontAwesomeStyles.fa,
        FontAwesomeStyles['fa-birthday-cake'],
        FontAwesomeStyles['fa-4x'],
      ]}
    />
  ))
  .add('with screenreader text', () => (
    <Icon
      id="SampleIcon"
      className={[
        FontAwesomeStyles.fa,
        FontAwesomeStyles['fa-smile-o'],
        FontAwesomeStyles['fa-4x'],
      ]}
      screenReaderText="Happy Icon"
    />
  ));
