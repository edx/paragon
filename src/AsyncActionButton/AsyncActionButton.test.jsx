import React from 'react';
import renderer from 'react-test-renderer';

import AsyncActionButton from './index';
import Icon from '../Icon';

describe('AsyncActionButton', () => {
  it('renders basic usage', () => {
    const props = {
      labels: {
        default: 'Saved',
        pending: 'Saving',
        complete: 'Saved',
      },
      className: 'btn-primary mr-2',
    };
    const tree = renderer.create((
      <div>
        <AsyncActionButton {...props} />
        <AsyncActionButton state="pending" {...props} />
        <AsyncActionButton state="complete" {...props} />
      </div>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders custom icons and disable during state', () => {
    const downloadButtonProps = {
      labels: {
        default: 'Download',
        pending: 'Downloading',
        complete: 'Downloaded',
      },
      icons: {
        default: <Icon className="fa fa-download" />,
        pending: <Icon className="fa fa-spinner fa-spin" />,
        complete: <Icon className="fa fa-check" />,
      },
      disabledStates: ['pending'],
      className: 'btn-primary mr-2',
    };
    const tree = renderer.create((
      <React.Fragment>
        <AsyncActionButton state="default" {...downloadButtonProps} />
        <AsyncActionButton state="pending" {...downloadButtonProps} />
        <AsyncActionButton state="complete" {...downloadButtonProps} />
      </React.Fragment>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders custom states', () => {
    const buttonProps = {
      labels: {
        unedited: 'Save (no changes)',
        edited: 'Save Changes',
      },
      icons: {
        unedited: <Icon className="fa fa-save" />,
        edited: <Icon className="fa fa-save" />,
      },
      disabledStates: ['unedited'],
      className: 'btn-primary mr-2',
    };
    const tree = renderer.create((
      <React.Fragment>
        <AsyncActionButton state="unedited" {...buttonProps} />
        <AsyncActionButton state="edited" {...buttonProps} />
      </React.Fragment>
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
