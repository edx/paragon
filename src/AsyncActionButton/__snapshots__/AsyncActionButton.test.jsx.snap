// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncActionButton renders basic usage 1`] = `
<div>
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-default btn-primary mr-2"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      Saved
    </span>
  </button>
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-pending btn-primary mr-2 disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="icon fa fa-spinner fa-spin"
          id="Icon1"
        />
      </span>
      Saving
    </span>
  </button>
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-complete btn-primary mr-2 disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="icon fa fa-check-circle"
          id="Icon1"
        />
      </span>
      Saved
    </span>
  </button>
</div>
`;

exports[`AsyncActionButton renders custom icons and disable during state 1`] = `
Array [
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-default btn-primary mr-2"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="fa fa-download"
          id="Icon1"
        />
      </span>
      Download
    </span>
  </button>,
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-pending btn-primary mr-2 disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="fa fa-spinner fa-spin"
          id="Icon1"
        />
      </span>
      Downloading
    </span>
  </button>,
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-complete btn-primary mr-2"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="fa fa-check"
          id="Icon1"
        />
      </span>
      Downloaded
    </span>
  </button>,
]
`;

exports[`AsyncActionButton renders custom states 1`] = `
Array [
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-unedited btn-primary mr-2 disabled"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="fa fa-save"
          id="Icon1"
        />
      </span>
      Save (no changes)
    </span>
  </button>,
  <button
    aria-live="assertive"
    className="btn pgn__async-action-btn pgn__async-action-btn-state-edited btn-primary mr-2"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="d-flex align-items-center justify-content-center"
    >
      <span
        className="pgn__async-action-btn-icon"
      >
        <span
          aria-hidden={true}
          className="fa fa-save"
          id="Icon1"
        />
      </span>
      Save Changes
    </span>
  </button>,
]
`;
